src/app/ui/timeline.cpp

```diff
@@ -1518,6 +1553,104 @@ void Timeline::drawCel(ui::Graphics* g, LayerIndex layerIndex, frame_t frame, Ce
   // Draw decorators to link the activeCel with its links.
   if (data->activeIt != data->end)
     drawCelLinkDecorators(g, bounds, cel, frame, is_active, is_hover, data);
+
+
+  if(m_celPreview && image) {
+
+    she::Surface *thumb_surf;
+
+#ifdef CEL_THUMB_CACHE
+    // FIXME limit memory usage
+    // TODO clean cache of closed document
+    doc::ObjectId image_id = image->id();
+    ThumbCache::iterator it;
+    if(is_active) {
+      m_celPreviewThumbCache.erase(image_id);
+      it = m_celPreviewThumbCache.end();
+    } else {
+      it = m_celPreviewThumbCache.find(image_id);
+    }
+    if(it != m_celPreviewThumbCache.end()) {
+      thumb_surf = (*it).second.get();
+    } else {
+#endif
+
+      base::UniquePtr<Image> thumb_img(Image::create(
+        image->pixelFormat(), bounds.w, bounds.h));
+
+      algorithm::scale_image(thumb_img, image,
+                             0, 0, thumb_img->width(), thumb_img->height(),
+                             0, 0, image->width(),     image->height());
+
+      thumb_surf = she::instance()->createRgbaSurface(
+        thumb_img->width(),
+        thumb_img->height());
+
+      convert_image_to_surface(thumb_img, m_sprite->palette(m_frame), thumb_surf,
+        0, 0, 0, 0, thumb_img->width(), thumb_img->height());
+
+#ifdef CEL_THUMB_CACHE
+      m_celPreviewThumbCache[image_id] = base::SharedPtr<she::Surface>(thumb_surf);
```

Comment:
`m_celPreviewThumbCache` contains `SharedPtr` but they are deleted with `delete`, and a `she::Surface` should be deleted with its `dispose` member.